src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp

```diff
@@ -280,56 +277,36 @@ real64 CompositionalMultiphaseFVM::scalingForSystemSolution( DomainPartition con
-          }
-
-          // compute the change in component densities and component fractions
-          for( localIndex ic = 0; ic < NC; ++ic )
-          {
-            localIndex const lid = dofNumber[ei] + ic + 1 - rankOffset;
-
-            // compute scaling factor based on relative change in component densities
-            real64 const absCompDensChange = fabs( localSolution[lid] );
-            real64 const maxAbsCompDensChange = maxCompFracChange * prevTotalDens;
-
-            // This actually checks the change in component fraction, using a lagged total density
-            // Indeed we can rewrite the following check as:
-            //    | prevCompDens / prevTotalDens - newCompDens / prevTotalDens | > maxCompFracChange
-            // Note that the total density in the second term is lagged (i.e, we use prevTotalDens)
-            // because I found it more robust than using directly newTotalDens (which can vary also
-            // wildly when the compDens change is large)
-            if( absCompDensChange > maxAbsCompDensChange && absCompDensChange > eps )
-            {
-              minVal.min( maxAbsCompDensChange / absCompDensChange );
-            }
-          }
-        }
-      } );
+      real64 const subRegionScalingFactor =
+        m_isThermal
+  ? thermalCompositionalMultiphaseBaseKernels::
+          ScalingForSystemSolutionKernelFactory::
+          createAndLaunch< parallelDevicePolicy<> >( m_maxRelativePresChange,
+                                                     m_maxRelativeTempChange,
+                                                     m_maxCompFracChange,
+                                                     dofManager.rankOffset(),
+                                                     m_numComponents,
+                                                     dofKey,
+                                                     subRegion,
+                                                     localSolution )
+  : isothermalCompositionalMultiphaseBaseKernels::
+          ScalingForSystemSolutionKernelFactory::
+          createAndLaunch< parallelDevicePolicy<> >( m_maxRelativePresChange,
+                                                     m_maxCompFracChange,
+                                                     dofManager.rankOffset(),
+                                                     m_numComponents,
+                                                     dofKey,
+                                                     subRegion,
+                                                     localSolution );
 
-      if( minVal.get() < scalingFactor )
+      if( subRegionScalingFactor < scalingFactor )
       {
-        scalingFactor = minVal.get();
+        scalingFactor = subRegionScalingFactor;
       }
     } );
   } );
 
-  return LvArray::math::max( MpiWrapper::min( scalingFactor, MPI_COMM_GEOSX ), m_minScalingFactor );
+  return LvArray::math::max( MpiWrapper::min( scalingFactor ), m_minScalingFactor );
 }
```

Comment:
I am a little bit reluctant to putting the MPI reduction in the `subRegion` loop because in some cases we have many `subRegions` and this may cause a lot of communications. That is why I always put the MPI reduction at the end of the function, outside the loops, even if it is not very pretty...